### DISCUSSION TITLE
How to implement a HUD based on a Billboard node?

### DISCUSSION THREAD

**Discussion by amrytov** [cyberbotics/webots#5424] [Q&A]
Hello,

The Billboard node would be, supposedly, very useful to implement "head-up displays" that are fixed relative to the viewpoint, precisely controllable and can be made visually interesting (much more so than the usual overlay panels).  To use it so, we must be able to place a display surface inside the Billboard somehow.

However, nor Robot nor device nodes can be placed inside a Billboard,  Billboard can't be placed inside a Robot.  The only workaround I could think of was to link a Display node via the DEF/REF mechanism to a texture in a Shape inside Billboard.  However, I couldn't make this work, may be I have been doing something wrong.  But without a way to connect to a Display node somehow, the Billboard nodes look rather useless. 

Have somebody been able to use Billboard with Displays somehow?  Is there a working demo?

**Comment by stefaniapedrazzi** [comment]
Hi,

the basic idea is, how you also pointed out, to add a plane or box with a texture where you draw the information you want to display.
You don't need a Display node to write/draw on the texture but you can use some external languages for do it.
For example if you are using Python there is the PIL module that has provides this functionality.

To inform Webots that has to update the texture you could simply use a stack of two (or more) images and rotate them using a Supervisor controller program to change the URL in the appearance node.

Would this approach work for your application?

**Reply by amrytov** [reply]
Hi, Stefania,

I admit that writing every change in the displayed data (such as seconds counting, for example) to a file (maybe even on a remote server) is a possible workaround, although probably not the most efficient one.

However, there is already a simple but in many cases sufficient mechnism of displaying data via the Display node,  which works totally in memory and (unless you specifically need to load images) doesn't involve any local or remote URL access.   It is also possible to process image data in code using PIL or other libraries, as you suggested, than copy it to a Display surface.   There is also a Billboard node which was specifically introduced to implement HUDs.  But it doesn't work with Displays, does it?   Seems to me, the most logical and efficient way would be to allow Display work with textured surfices within a Billboard?

**Reply by stefaniapedrazzi** [reply]
I agree.
The Billboard node would have much more potential but it is currently very limited.
We are planning to add some more functionalities in the future.
We will take your suggestions into consideration to evaluate what is the best approach to improve the Billboard/Display functionality.
